src/main/scala/uk/gov/nationalarchives/omega/api/services/ApiService.scala

```diff
@@ -0,0 +1,146 @@
+
+  private def attemptStartUp(): IO[ExitCode] =
+    // switch to the started state
+    switchState(Starting, Started)
+      .ifM(
+        doStart(),
+        IO.pure(ExitCode(invalidState))
+      )
+      .handleErrorWith(error =>
+        logger.error(s"Shutting down due to ${error.getMessage}") *>
+          doStop() *> switchState(Starting, Stopped) *> IO.pure(ExitCode.Error)
+      )
+
+  def stop(): IO[ExitCode] =
+    // check if we can stop the service
+    switchState(Started, Stopping).ifM(
+      logger.info("Closing connection..") *>
+        switchState(Stopping, Stopped).ifM(doStop(), IO.pure(ExitCode(invalidState))),
+      IO.pure(ExitCode(invalidState))
+    )
+
+  private def doStart(): IO[ExitCode] = {
+
+    // TODO(AR) - one client, how to ack a consumer message after local persistence and then process it, and then produce a response
+    // TODO(AR) how to wire up queues and services using a config file or DSL?
+    // TODO(AR) request queue will typically be 1 (plus maybe a few more for expedited ops), response queues will be per external application
+    val localQueue: IO[Queue[IO, LocalMessage]] = Queue.bounded[IO, LocalMessage](config.maxLocalQueueSize)
+    val jmsConnector = new JmsConnector(config)
+    getLocalMessageStore.flatMap {
+      case Right(localMessageStore) =>
+        val result = for {
+          q <- localQueue
+          res <-
+            jmsConnector.getJmsProducerAndConsumer(QueueName(config.requestQueue)).use {
+              case (jmsProducer, jmsConsumer) =>
+                process(q, jmsConsumer, jmsProducer, localMessageStore)
+            }
+        } yield res
+        result.as(ExitCode.Success)
+      case Left(exitCode) => IO.pure(exitCode)
+    }
+  }
+
+  private def getLocalMessageStore: IO[Either[ExitCode, LocalMessageStore]] = IO {
+    val messageStoreFolder = Paths.get(config.tempMessageDir)
+    Files.createDirectories(messageStoreFolder)
+    Right(new LocalMessageStore(messageStoreFolder))
+  }.handleErrorWith(ex =>
+    logger.error(s"Failed to created local message store due to ${ex.getMessage}") *> IO.pure(Left(ExitCode.Error))
+  )
+
+  private def process(
+    queue: Queue[IO, LocalMessage],
+    consumer: JmsAcknowledgerConsumer[IO],
+    producer: JmsProducer[IO],
+    localMessageStore: LocalMessageStore
+  ) =
+    IO.race(
+      createMessageHandler(queue, localMessageStore)(consumer),
+      List.range(start = 0, end = config.maxDispatchers).parTraverse_ { i =>
+        logger.info(s"Starting consumer #${i + 1}") >>
+          new Dispatcher(new LocalProducerImpl(producer, QueueName(config.replyQueue)), new EchoService())
+            .run(i)(queue)
+            .foreverM
+      }
+    )
+
+  private def doStop(): IO[ExitCode] =
+    // TODO(RW) this is where we will need to close any external connections, for example to OpenSearch
+    logger.info("Connection closed.") *> IO.pure(ExitCode.Success)
+
+  private def acknowledgeMessage(): IO[AckAction[IO]] =
```

Comment:
We use a `JmsAcknowledgerConsumer` from the jms4s library rather than a `JmsAutoAcknowledgerConsumer` so that we can hold back acknowledgement of the message until we are sure a copy is safely stored on disk. This function just triggers the acknowledgement.